### PR TITLE
gitignore with escapes

### PR DESCRIPTION
**Pull request by pks-t**
Various fixes to how we handle escapes when parsing gitignore patterns.

### CODE REVIEW

**Comment by ethomson** [comment]
Thanks for 65ad6b5afdc5df17a706ef36767bffcbc8c9ab55.  😢

**Comment by ethomson** [comment]
> As there is no platform that regards "\foo" as absolute, to the best of my knowledge, only treat "\" as absolute when we have a drive or Windows-style network path prefix. Add tests to catch regressions.

So this is an interesting question.  Is a path prefixed by a `\` absolute on Windows?  And I think the answer is yes.  Windows has the notion of a "current drive" that makes the drive specifier unnecessary in paths.  eg, if your current drive is `C:` and you want to open `\autoexec.bat` then this is implicitly `C:\autoexec.bat`.

You can see this on the shell.  If you're in cmd, you should be able to run `C:` to set your current drive to `C:`, then you should be able to run `type \autoexec.bat` and have it type `C:\autoexec.bat`.  (OK, you probably _can't_ do this, I'm not sure if `C:\autoexec.bat` exists on a modern system.  Replace with a path that you know to exist on your drive.)

What I'm not sure is what that means exactly for us.  I think that `CreateFile("\foo")` will absolutely take the current drive into account, so I think that it means that a `\` prefixed path is absolute on Windows.

I think that this is what the existing code was trying to communicate - and that it is correct, but only on Windows.

**Comment by ethomson** [comment]
I'm +1 on this otherwise.

**Comment by pks-t** [comment]
> I think that this is what the existing code was trying to communicate - and that it is correct, but only on Windows.

Okay, thanks for clarifying. My first approach was to simply move that code into the Windows-specific `ifdef` section. But I've been mighty surprised by us treating "C:\"-style drive prefixes as absolute on all systems, but network-style absolute paths are only treated as absolute on Windows. So... I guess we should change that to also treat "C:\"-style things as absolute on Windows-based systems, only?

**Comment by ethomson** [comment]
>  Okay, thanks for clarifying. My first approach was to simply move that code into the Windows-specific ifdef section. But I've been mighty surprised by us treating "C:"-style drive prefixes as absolute on all systems, but network-style absolute paths are only treated as absolute on Windows. So... I guess we should change that to also treat "C:"-style things as absolute on Windows-based systems, only?

Ugh.  I _think_ so?  Importantly, we (try to) do this right in the network code, where `C:/foo` is a local path on Windows and an ssh path everywhere else.

But that's a legitimate _relative_ path on (non-macOS) Unix systems, so...  yeah, we should probably be treating it as absolute _only_ on Windows systems.

**Comment by ethomson** [comment]
> But that's a legitimate relative path on (non-macOS) Unix systems

Also, this is a lie.  It's a legitimate path on macOS as well.  I thought this was actually a reserved character in the Mac filesystem.  (For pre-OS X compatibility, which used `:` as a path delimiter.)

But I'm mistaken.  Some apps still don't allow you to create files with `:` in them (notably, Finder) but this is supported in the filesystem.

File: src/attr_file.c [hunk 1]
```diff
@@ -587,6 +587,7 @@ int git_attr_fnmatch__parse(
 {
 	const char *pattern, *scan;
 	int slash_count, allow_space;
+	char escaped;
```
[the remark this answers is not in the document]
Comment (pks-t) [reply src/attr_file.c]: Fixed here and in `unescape_spaces`.

**Comment by pks-t** [comment]
So I've changed `git_path_root` now to treat paths with leading backslash as absolute on Windows-based systems, only. All the other systems will now treat those as relative paths. I've refrained from modifying existing behaviour with drive-prefixes, I'd rather do that in a separate PR.

**Comment by ethomson** [comment]
Sweet, thanks!

**Pull request merged**